### PR TITLE
refactor Flux transformer to use scanned blocks, dynamic checkpointing, and decoupled projections

### PR DESCRIPTION
**Pull request by prishajain1**
### Overview 
This PR refactors the Flux model architecture in MaxDiffusion to support scanned blocks (nn.scan) for double and single blocks, implements configurable gradient checkpointing (rematerialization) policies, and updates the weights loader to support loading pretrained checkpoints under the scanned format.

### Key Changes

- Decoupled Fused Projections: Decoupled the projection layers (implementing the `MlpAndOutputBlock` wrapper) to eliminate redundant recomputation of attention and projection outputs.
- QKV Slicing Refactoring: Refactored the QKV projection slicing logic to use `jnp.split` across Flux transformer blocks for cleaner layout constraints.
- Scanned Block Architecture: Migrated Flux Double and Single Transformer Blocks to use `nn.scan` to optimize compiler tracing and step execution speed on TPUs.
- Dynamic Gradient Checkpointing: Added `FLUX_OPTIMIZED` to `GradientCheckpointType` to allow configuring block-specific rematerialization policies dynamically via configuration files instead of being hardcoded.
- Stacked Weights Loading: Updated the weights loader (`util.py`) to slice, group, and stack PyTorch checkpoint weights along axis 0 to match the expected format of `nn.scan` layers.

### CODE REVIEW

File: src/maxdiffusion/checkpointing/flux_checkpointer.py [hunk 1]
```diff
@@ -214,6 +214,11 @@ def load_diffusers_checkpoint(self):
           dtype=self.config.activations_dtype,
           weights_dtype=self.config.weights_dtype,
           precision=max_utils.get_precision(self.config),
+          use_base2_exp=self.config.use_base2_exp,
+          use_experimental_scheduler=self.config.use_experimental_scheduler,
+          remat_policy=self.config.remat_policy,
+          names_which_can_be_saved=self.config.names_which_can_be_saved,
+          names_which_can_be_offloaded=self.config.names_which_can_be_offloaded,
```
Comment (Perseus14) [inline src/maxdiffusion/checkpointing/flux_checkpointer.py:221]: Can we make these variable names a little more explicit? Maybe something like `saved_transformer_layer_names` or `savable_transformer_layer_names` and `offloaded_transformer_layer_names` or `offloadable_transformer_layer_names`? I will leave it upto you

File: src/maxdiffusion/checkpointing/flux_checkpointer.py [hunk 2]
```diff
@@ -279,6 +284,11 @@ def load_checkpoint(self, step=None, scheduler_class=None):
             weights_dtype=self.config.weights_dtype,
             precision=max_utils.get_precision(self.config),
             from_pt=self.config.from_pt,
+            use_base2_exp=self.config.use_base2_exp,
+            use_experimental_scheduler=self.config.use_experimental_scheduler,
+            remat_policy=self.config.remat_policy,
+            names_which_can_be_saved=self.config.names_which_can_be_saved,
+            names_which_can_be_offloaded=self.config.names_which_can_be_offloaded,
         )
```
Comment (Perseus14) [inline src/maxdiffusion/checkpointing/flux_checkpointer.py:292]: Same as above

File: src/maxdiffusion/configs/base_flux_dev.yml [hunk 3]
```diff
@@ -218,6 +222,21 @@ transform_images_num_proc: 4
...
+# MATMUL_WITHOUT_BATCH - means gradient checkpoint for every linear/matmul operation,
+# except for ones that involve batch dimension - that means that all attention and projection
+# layers will have gradient checkpoint, but not the backward with respect to the parameters.
+# OFFLOAD_MATMUL_WITHOUT_BATCH - same as MATMUL_WITHOUT_BATCH but offload instead of recomputing.
+# CUSTOM - set names to offload and save.
+remat_policy: "FLUX_OPTIMIZED"
+# For CUSTOM policy set below, current annotations are for: attn_output, query_proj, key_proj, value_proj
+# xq_out, xk_out, ffn_activation
+names_which_can_be_saved: []
+names_which_can_be_offloaded: []
```
Comment (Perseus14) [inline src/maxdiffusion/configs/base_flux_dev.yml:237]: Same as above

File: src/maxdiffusion/models/flux/transformers/transformer_flux_flax.py
Comment (Perseus14) [inline src/maxdiffusion/models/flux/transformers/transformer_flux_flax.py:1]: Can we rename the file to `transformer_flux.py`?

File: src/maxdiffusion/models/flux/transformers/transformer_flux_flax.py [hunk 4]
```diff
@@ -138,37 +188,22 @@ def __call__(self, hidden_states, temb, image_rotary_emb=None):
...
-  Reference: https://arxiv.org/abs/2403.03206
-
-  Parameters:
-      dim (`int`): The number of channels in the input and output.
-      num_attention_heads (`int`): The number of heads to use for multi-head attention.
-      attention_head_dim (`int`): The number of channels in each head.
-      context_pre_only (`bool`): Boolean to determine if we should add some blocks associated with the
-          processing of `context` conditions.
-  """
-
```
Comment (Perseus14) [inline src/maxdiffusion/models/flux/transformers/transformer_flux_flax.py:166]: Let's keep the comment block. Update if required

File: src/maxdiffusion/models/flux/transformers/transformer_flux_flax.py [hunk 5]
```diff
@@ -341,7 +395,12 @@ class FluxTransformer2DModel(nn.Module, FlaxModelMixin, ConfigMixin):
   qkv_bias: bool = True
   theta: int = 1000
   attention_kernel: str = "dot_product"
-  eps = 1e-6
+  eps: float = 1e-6
+  remat_policy: str = "None"
+  names_which_can_be_saved: tuple = ()
+  names_which_can_be_offloaded: tuple = ()
```
Comment (Perseus14) [inline src/maxdiffusion/models/flux/transformers/transformer_flux_flax.py:413]: Please change variable name

File: src/maxdiffusion/models/flux/transformers/transformer_flux.py [hunk 6]
```diff
@@ -532,24 +633,23 @@ def init_weights(self, rngs, max_sequence_length, eval_only=True):
     batch_size = 1 * num_devices
     batch_image_shape = (
         batch_size,
-        16,  # 16 to match jflux.get_noise
+        16,
```
Comment (Perseus14) [inline src/maxdiffusion/models/flux/transformers/transformer_flux.py]: Can we move these hardcoded values to some variables?

File: src/maxdiffusion/models/flux/transformers/transformer_flux.py [hunk 7]
```diff
@@ -532,24 +633,23 @@ def init_weights(self, rngs, max_sequence_length, eval_only=True):
...
         2 * resolution // scale_factor,
         2 * resolution // scale_factor,
     )
-    # bs, encoder_input, seq_length
     text_shape = (
         batch_size,
         max_sequence_length,
-        4096,  # Sequence length of text encoder, how to get this programmatically?
+        4096,
```
Comment (Perseus14) [inline src/maxdiffusion/models/flux/transformers/transformer_flux.py]: Same as above

File: src/maxdiffusion/models/flux/transformers/transformer_flux.py [hunk 8]
```diff
@@ -532,24 +633,23 @@ def init_weights(self, rngs, max_sequence_length, eval_only=True):
...
     )
     text_ids_shape = (
         batch_size,
         max_sequence_length,
-        3,  # Hardcoded to match jflux.prepare
+        3,
```
Comment (Perseus14) [inline src/maxdiffusion/models/flux/transformers/transformer_flux.py]: Same as above

File: src/maxdiffusion/models/flux/transformers/transformer_flux.py [hunk 9]
```diff
@@ -532,24 +633,23 @@ def init_weights(self, rngs, max_sequence_length, eval_only=True):
...
     )
     vec_shape = (
         batch_size,
-        768,  # Sequence length of clip, how to get this programmatically?
+        768,
```
Comment (Perseus14) [inline src/maxdiffusion/models/flux/transformers/transformer_flux.py]: Same as above

File: src/maxdiffusion/models/normalization_flax.py [hunk 10]
```diff
@@ -51,14 +51,6 @@ def __call__(self, x, conditioning_embedding):
...
 
 class AdaLayerNormZero(nn.Module):
-  r"""
-  Norm layer adaptive layer norm zero (adaLN-Zero).
-
-  Parameters:
-      embedding_dim (`int`): The size of each embedding vector.
-      num_embeddings (`int`): The size of the embeddings dictionary.
-  """
-
```
Comment (Perseus14) [inline src/maxdiffusion/models/normalization_flax.py:61]: Let's keep the comment block

File: src/maxdiffusion/models/normalization_flax.py [hunk 11]
```diff
@@ -77,38 +73,30 @@ def __call__(self, x, emb):
...
 
 class AdaLayerNormZeroSingle(nn.Module):
-  r"""
-  Norm layer adaptive layer norm zero (adaLN-Zero).
-
-  Parameters:
-      embedding_dim (`int`): The size of each embedding vector.
-      num_embeddings (`int`): The size of the embeddings dictionary.
-  """
-
```
Comment (Perseus14) [inline src/maxdiffusion/models/normalization_flax.py:111]: Let's keep the comment block

File: src/maxdiffusion/models/normalization_flax.py [hunk 12]
```diff
@@ -129,24 +119,27 @@ def __call__(self, x, emb):
...
-          + shift_msa
-      )
+      # Fused optimization math keeping exact pretrained weight compatibility
+      mean = jnp.mean(x, axis=-1, keepdims=True)
+      variance = jnp.mean(jnp.square(x - mean), axis=-1, keepdims=True)
+      inv_std = jax.lax.rsqrt(variance + 1e-6)
+
+      x = (x - mean) * inv_std * (1.0 + scale_msa) + shift_msa
     else:
-      raise ValueError(f"Unsupported `norm_type` ({self.norm_type}) provided. Supported ones are: 'layer_norm'.")
```
Comment (Perseus14) [inline src/maxdiffusion/models/normalization_flax.py:149]: Is there a need to change this text comment?

File: src/maxdiffusion/generate_flux.py [hunk 13]
```diff
@@ -314,6 +314,9 @@ def run(config):
         dtype=config.activations_dtype,
         weights_dtype=config.weights_dtype,
         precision=get_precision(config),
+        remat_policy=config.remat_policy,
+        names_which_can_be_saved=config.names_which_can_be_saved,
+        names_which_can_be_offloaded=config.names_which_can_be_offloaded,
```
Comment (Perseus14) [inline src/maxdiffusion/generate_flux.py:319]: Please change variable names

**Review by Perseus14 — CHANGES_REQUESTED** [review]

File: src/maxdiffusion/checkpointing/flux_checkpointer.py (hunk 1 above)
Comment (prishajain1) [reply src/maxdiffusion/checkpointing/flux_checkpointer.py:221]: Other models such as WAN and LTX already use these exact parameter names, I suggest we do not change it to maintain consistency across the codebase

File: src/maxdiffusion/models/flux/transformers/transformer_flux_flax.py
Comment (prishajain1) [reply src/maxdiffusion/models/flux/transformers/transformer_flux_flax.py:1]: Done

File: src/maxdiffusion/models/flux/transformers/transformer_flux_flax.py (hunk 4 above)
Comment (prishajain1) [reply src/maxdiffusion/models/flux/transformers/transformer_flux_flax.py:166]: Done

File: src/maxdiffusion/models/flux/transformers/transformer_flux.py (hunk 6 above)
Comment (prishajain1) [reply src/maxdiffusion/models/flux/transformers/transformer_flux.py]: Done!

File: src/maxdiffusion/models/normalization_flax.py (hunk 12 above)
Comment (prishajain1) [reply src/maxdiffusion/models/normalization_flax.py:149]: Reverted

**Review by Perseus14 — DISMISSED** [review]

File: src/maxdiffusion/checkpointing/flux_checkpointer.py (hunk 1 above)
Comment (entrpn) [reply src/maxdiffusion/checkpointing/flux_checkpointer.py:221]: yes lets keep consistency.

File: src/maxdiffusion/configs/base_flux_dev.yml [hunk 14]
```diff
@@ -147,9 +149,11 @@ mesh_axes: ['data', 'fsdp', 'context', 'tensor']
 # conv_in : conv.shape[2] weight
 # conv_out : conv.shape[-1] weight
 logical_axis_rules: [
-                      ['batch', 'data'],
+                      ['batch', ['data','fsdp']],
                       ['activation_batch', ['data','fsdp']],
-                      ['activation_heads', 'tensor'],
+                      ['activation_heads', 'fsdp'],
```
Comment (entrpn) [inline src/maxdiffusion/configs/base_flux_dev.yml]: shouldn't activation_heads be sharded across tensor? Or is that Wan only?

**Review by entrpn — COMMENTED** [review]
@prishajain1 did you update the readme with the best values based on remat?

File: src/maxdiffusion/configs/base_flux_dev.yml (hunk 14 above)
Comment (prishajain1) [reply src/maxdiffusion/configs/base_flux_dev.yml]: Updated. I have tested on v5p to see if changing this leads to any OOMs or latency regression and didn't find any such case

**Comment by prishajain1** [comment]
> @prishajain1 did you update the readme with the best values based on remat?

Updated

**Review by Perseus14 — DISMISSED** [review]

**Review by entrpn — DISMISSED** [review]
